abstract_environmentt::eval rework

jezhiggins opened the pull request:
Refactoring work centred on abstract_environmentt::eval, trying to reduce the volume of special case handling. 

Moved implementation of indexing, member access, and pointer dereferencing out into the target object types. This allowed us to combine several of the if-ladder cases together.

jezhiggins committed:
index expression unit test

jezhiggins committed:
expand out array index unit test

jezhiggins committed:
Remove redundant value_set_array_abstract_object overrides

jezhiggins committed:
Move index_expression evaluation into array_abstract_object

Move it out of abstract_environment.

jezhiggins committed:
Very simple union member access test

union_abstract_object is only top or bottom, consequently this test
tells us very little beyond telling us union member access works doesn't
explode. However there is nowhere else in the tests that  doing anything
at all with unions, so it is a least a little backstop.

jezhiggins committed:
Move member expression evaluation into struct_abstract_object

Move it out of abstract_environment. union_abstract_objects can also
be the target of member expressions, but this is not implemented in the
current code so I've left it out of here too.

jezhiggins committed:
Combine member_expr and index_expr cases together

jezhiggins committed:
Removed redundant value_set_pointer_abstract_object member functions

jezhiggins committed:
Move dereference evaluation into pointer_abstract_object

Move it out of abstract_environment.

jezhiggins committed:
Remove redundant cast

jezhiggins committed:
Combine array, struct, and constant handlers

jezhiggins committed:
Factored out eval_operands helper function

jezhiggins committed:
Clang format

martin-cs commented on regression/goto-analyzer/simple-union-access/test.desc:
```diff
@@ -0,0 +1,15 @@
+CORE
+union.c
+--variable-sensitivity --vsd --verify
```
I think `--vsd` and `--variable-sensitivity` should do the same thing.  One was for backwards compatibility.

martin-cs replied:
Please can you go down to just one of these two?

martin-cs commented on regression/goto-analyzer/simple-union-access/test.desc:
```diff
@@ -0,0 +1,15 @@
+^\[main.assertion.5\] line 19 x.a==0: UNKNOWN$
+^\[main.assertion.6\] line 20 x.a==1: UNKNOWN$
+^\[main.assertion.7\] line 21 x.b==0: UNKNOWN$
+^\[main.assertion.8\] line 22 x.b==1: UNKNOWN$
```
With a "most recently written" abstraction for unions it should be possible to turn about half of these into successes but for now all unknown is the best that can be done.

martin-cs commented on unit/analyses/variable-sensitivity/eval-member-access.cpp:
```diff
@@ -0,0 +1,111 @@
+  namespacet &ns
+);
+
+SCENARIO(
```
Unit tests are good.  In theory much of VSD should be testable via unit tests so having more is a good thing.

martin-cs commented on unit/analyses/variable-sensitivity/eval-member-access.cpp:
```diff
@@ -0,0 +1,158 @@
+  auto top_value =
+    std::make_shared<abstract_valuet>(integer_typet(), true, false);
+  return top_value->to_constant();
+}
```
Missing new line.  Only worth fixing if you have other things to fix.

martin-cs commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -68,35 +74,12 @@ abstract_environmentt::eval(const exprt &expr, const namespacet &ns) const
-    return array_abstract_object->read(*this, index_expr, ns);
-  }
-  else if(simplified_id == ID_array)
+  else if(
```
Yes; these are all basically the same constant case.

martin-cs commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -51,14 +56,15 @@ abstract_environmentt::eval(const exprt &expr, const namespacet &ns) const
     }
   }
-  else if(simplified_id == ID_member)
+  else if(
```
Can this case be folded into the general `else` case?

martin-cs commented on src/analyses/variable-sensitivity/array_abstract_object.cpp:
```diff
@@ -35,6 +35,21 @@ array_abstract_objectt::array_abstract_objectt(
+{
+  if(expr.id() == ID_index)
+  {
+    return read(environment, expr, ns);
```
Given that `read` just calls `read_index` can you call `read_index` directly here (and the corresponding functions in struct, pointer, etc.) and then remove `read` from the interface of `abstract_objectt` completely.

martin-cs reviewed (changes requested):
Yes; this is directly in the right direction and does simplify things.

Request changes because it would be good if it could go a little bit further and remove `abstract_objectt::read` completely by calling the internal functions `read_*` of the relevant classes.

martin-cs commented:
Reflecting on this, I think the only uses of `read_index`, `read_component`, etc. are in `read` so you could inline the contents of these into the `expression_transform` and remove everything `read` related.

jezhiggins committed:
Remove abstract_objectt::read member function

Now we've moved some work into expression_transform, we don't need a
read member on the base class.

martin-cs commented:
Better -- thank you.

jezhiggins committed:
Add eval-expression-transform to unit tests

martin-cs reviewed (approved):
One tweak to the union regression test and should be good to merge.

jezhiggins committed:
Remove redundant command line options
